File: src/mobile-client/android/LivenessClient.kt
/**
 * Android Client SDK for Liveness Detection Pipeline
 *
 * Integrates with the Azure-based liveness detection pipeline.
 * Uses Azure AI Vision Face SDK for client-side liveness detection.
 *
 * Prerequisites:
 * - Azure AI Vision Face SDK
 * - Kotlin Coroutines
 * - OkHttp / Retrofit (recommended)
 * - Minimum API Level 24
 */

package com.liveness.pipeline.client

import android.graphics.Bitmap
import android.provider.Settings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.TimeUnit

// MARK: - Models

/**
 * Pipeline session response from the server
 */
@Serializable
data class PipelineSession(
    @SerialName("session_id") val sessionId: String,
    @SerialName("status") val status: String,
    @SerialName("token") val token: String,
    @SerialName("token_expires_at") val tokenExpiresAt: String,
    @SerialName("liveness_result") val livenessResult: String? = null,
    @SerialName("liveness_confidence") val livenessConfidence: Double? = null,
    @SerialName("applicant_id") val applicantId: String? = null,
    @SerialName("error") val error: String? = null
)

/**
 * Azure liveness session for SDK integration
 */
@Serializable
data class AzureLivenessSession(
    @SerialName("azure_session_id") val azureSessionId: String,
    @SerialName("auth_token") val authToken: String,
    @SerialName("pipeline_session_id") val pipelineSessionId: String,
    @SerialName("pipeline_token") val pipelineToken: String
)

/**
 * Applicant data for submission
 */
@Serializable
data class ApplicantData(
    @SerialName("first_name") val firstName: String? = null,
    @SerialName("last_name") val lastName: String? = null,
    @SerialName("email") val email: String? = null,
    @SerialName("phone") val phone: String? = null,
    @SerialName("date_of_birth") val dateOfBirth: String? = null,
    @SerialName("address") val address: String? = null,
    @SerialName("custom_fields") val customFields: Map<String, String>? = null
)

/**
 * Session creation request
 */
@Serializable
data class CreateSessionRequest(
    @SerialName("client_id") val clientId: String,
    @SerialName("applicant_data") val applicantData: ApplicantData,
    @SerialName("callback_url") val callbackUrl: String? = null,
    @SerialName("metadata") val metadata: Map<String, String>? = null
)

/**
 * Liveness detection result
 */
enum class LivenessResult(val value: String) {
    LIVE("live"),
    SPOOF("spoof"),
    UNCERTAIN("uncertain"),
    UNKNOWN("unknown");

    companion object {
        fun fromString(value: String?): LivenessResult {
            return when (value?.lowercase()) {
                "live" -> LIVE
                "spoof" -> SPOOF
                "uncertain" -> UNCERTAIN
                else -> UNKNOWN
            }
        }
    }
}

/**
 * Pipeline completion result
 */
data class PipelineResult(
    val sessionId: String,
    val livenessResult: LivenessResult,
    val livenessConfidence: Double,
    val applicantId: String?,
    val isSuccess: Boolean,
    val error: String?
)

// MARK: - Errors

/**
 * Errors that can occur during liveness detection
 */
sealed class LivenessClientException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class InvalidConfiguration(message: String = "Invalid client configuration") : LivenessClientException(message)
    class NetworkError(cause: Throwable) : LivenessClientException("Network error: ${cause.message}", cause)
    class InvalidResponse(message: String = "Invalid server response") : LivenessClientException(message)
    class SessionExpired(message: String = "Session has expired") : LivenessClientException(message)
    class LivenessCheckFailed(reason: String) : LivenessClientException("Liveness check failed: $reason")
    class ServerError(message: String) : LivenessClientException("Server error: $message")
}

// MARK: - Client Configuration

/**
 * Configuration for the liveness client
 */
data class LivenessClientConfig(
    val baseUrl: String,
    val clientId: String,
    val functionKey: String? = null,
    val timeoutSeconds: Long = 30
)

// MARK: - Liveness Client

/**
 * Main client for interacting with the liveness detection pipeline
 *
 * Example usage:
 * ```kotlin
 * val client = LivenessClient(LivenessClientConfig(
 *     baseUrl = "https://your-function-app.azurewebsites.net",
 *     clientId = "your-client-id",
 *     functionKey = "your-function-key"
 * ))
 *
 * // Complete flow
 * val result = client.performLivenessCheck(
 *     applicantData = ApplicantData(
 *         firstName = "John",
 *         lastName = "Doe",
 *         email = "john@example.com"
 *     ),
 *     image = capturedBitmap
 * )
 *
 * if (result.isSuccess) {
 *     println("Applicant verified! ID: ${result.applicantId}")
 * }
 * ```
 */
class LivenessClient(private val config: LivenessClientConfig) {

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
    }

    /**
     * Current active session
     */
    var currentSession: PipelineSession? = null
        private set

    // MARK: - Session Management

    /**
     * Start a new liveness detection session
     */
    suspend fun startSession(
        applicantData: ApplicantData,
        callbackUrl: String? = null,
        metadata: Map<String, String>? = null
    ): PipelineSession = withContext(Dispatchers.IO) {
        val request = CreateSessionRequest(
            clientId = config.clientId,
            applicantData = applicantData,
            callbackUrl = callbackUrl,
            metadata = metadata
        )

        val session = post<PipelineSession>(
            path = "/api/sessions",
            body = json.encodeToString(request)
        )

        currentSession = session
        session
    }

    /**
     * Create Azure liveness session for SDK integration
     */
    suspend fun createLivenessSession(
        deviceCorrelationId: String? = null
    ): AzureLivenessSession = withContext(Dispatchers.IO) {
        val session = currentSession ?: throw LivenessClientException.InvalidConfiguration()

        val body = deviceCorrelationId?.let {
            json.encodeToString(mapOf("device_correlation_id" to it))
        } ?: "{}"

        post(
            path = "/api/sessions/${session.sessionId}/liveness",
            body = body,
            token = session.token
        )
    }

    /**
     * Complete the liveness flow after Azure SDK finishes
     */
    suspend fun completeLiveness(
        azureSessionId: String
    ): PipelineSession = withContext(Dispatchers.IO) {
        val session = currentSession ?: throw LivenessClientException.InvalidConfiguration()

        val body = json.encodeToString(mapOf("azure_session_id" to azureSessionId))

        val result = post<PipelineSession>(
            path = "/api/sessions/${session.sessionId}/complete",
            body = body,
            token = session.token
        )

        currentSession = result
        result
    }

    /**
     * Submit image for server-side liveness detection
     */
    suspend fun submitImage(image: Bitmap): PipelineSession = withContext(Dispatchers.IO) {
        val session = currentSession ?: throw LivenessClientException.InvalidConfiguration()

        val stream = ByteArrayOutputStream()
        image.compress(Bitmap.CompressFormat.JPEG, 90, stream)
        val imageData = stream.toByteArray()

        val result = postImage(
            path = "/api/sessions/${session.sessionId}/image",
            imageData = imageData,
            token = session.token
        )

        currentSession = result
        result
    }

    /**
     * Get current session status
     */
    suspend fun getSessionStatus(): PipelineSession = withContext(Dispatchers.IO) {
        val session = currentSession ?: throw LivenessClientException.InvalidConfiguration()

        val result = get<PipelineSession>(
            path = "/api/sessions/${session.sessionId}",
            token = session.token
        )

        currentSession = result
        result
    }

    /**
     * Cancel the current session
     */
    suspend fun cancelSession(): Unit = withContext(Dispatchers.IO) {
        val session = currentSession ?: return@withContext

        delete(
            path = "/api/sessions/${session.sessionId}",
            token = session.token
        )

        currentSession = null
    }

    // MARK: - Complete Flow Helper

    /**
     * Execute the complete liveness detection flow
     */
    suspend fun performLivenessCheck(
        applicantData: ApplicantData,
        image: Bitmap
    ): PipelineResult = withContext(Dispatchers.IO) {
        // Step 1: Start session
        startSession(applicantData)

        // Step 2: Submit image for liveness detection
        val result = submitImage(image)

        // Step 3: Build result
        val livenessResult = LivenessResult.fromString(result.livenessResult)

        PipelineResult(
            sessionId = result.sessionId,
            livenessResult = livenessResult,
            livenessConfidence = result.livenessConfidence ?: 0.0,
            applicantId = result.applicantId,
            isSuccess = livenessResult == LivenessResult.LIVE && result.applicantId != null,
            error = result.error
        )
    }

    // MARK: - Network Helpers

    private fun buildUrl(path: String): URL {
        val baseUrl = config.baseUrl.trimEnd('/')
        return URL(baseUrl + path)
    }

    private fun setupConnection(
        url: URL,
        method: String,
        token: String? = null
    ): HttpURLConnection {
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = method
        connection.connectTimeout = TimeUnit.SECONDS.toMillis(config.timeoutSeconds).toInt()
        connection.readTimeout = TimeUnit.SECONDS.toMillis(config.timeoutSeconds).toInt()

        config.functionKey?.let {
            connection.setRequestProperty("x-functions-key", it)
        }

        token?.let {
            connection.setRequestProperty("Authorization", "Bearer $it")
        }

        return connection
    }

    private inline fun <reified T> get(
        path: String,
        token: String? = null
    ): T {
        val url = buildUrl(path)
        val connection = setupConnection(url, "GET", token)

        return executeRequest(connection)
    }

    private inline fun <reified T> post(
        path: String,
        body: String,
        token: String? = null
    ): T {
        val url = buildUrl(path)
        val connection = setupConnection(url, "POST", token)
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json")

        connection.outputStream.use { outputStream ->
            outputStream.write(body.toByteArray(Charsets.UTF_8))
        }

        return executeRequest(connection)
    }

    private fun postImage(
        path: String,
        imageData: ByteArray,
        token: String? = null
    ): PipelineSession {
        val url = buildUrl(path)
        val connection = setupConnection(url, "POST", token)
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "image/jpeg")

        connection.outputStream.use { outputStream ->
            outputStream.write(imageData)
        }

        return executeRequest(connection)
    }

    private fun delete(
        path: String,
        token: String? = null
    ) {
        val url = buildUrl(path)
        val connection = setupConnection(url, "DELETE", token)

        try {
            val responseCode = connection.responseCode

            if (responseCode >= 400 && responseCode != 404) {
                throw LivenessClientException.ServerError("HTTP $responseCode")
            }
        } catch (e: LivenessClientException) {
            throw e
        } catch (e: Exception) {
            throw LivenessClientException.NetworkError(e)
        } finally {
            connection.disconnect()
        }
    }

    private inline fun <reified T> executeRequest(connection: HttpURLConnection): T {
        try {
            val responseCode = connection.responseCode

            if (responseCode == 401) {
                throw LivenessClientException.SessionExpired()
            }

            val responseBody = if (responseCode >= 400) {
                connection.errorStream?.bufferedReader()?.readText() ?: ""
            } else {
                connection.inputStream.bufferedReader().readText()
            }

            if (responseCode >= 400) {
                val errorMessage = try {
                    val errorMap = json.decodeFromString<Map<String, String>>(responseBody)
                    errorMap["error"] ?: "HTTP $responseCode"
                } catch (e: Exception) {
                    "HTTP $responseCode"
                }
                throw LivenessClientException.ServerError(errorMessage)
            }

            return json.decodeFromString(responseBody)
        } catch (e: LivenessClientException) {
            throw e
        } catch (e: IOException) {
            throw LivenessClientException.NetworkError(e)
        } catch (e: Exception) {
            throw LivenessClientException.InvalidResponse("Failed to parse response: ${e.message}")
        } finally {
            connection.disconnect()
        }
    }
}

// MARK: - Azure SDK Integration

/**
 * Extension for Azure Face SDK integration
 *
 * Example usage with Azure Face SDK:
 * ```kotlin
 * val client = LivenessClient(config)
 *
 * // Start session
 * val session = client.startSession(applicantData)
 *
 * // Get Azure liveness session
 * val azureSession = client.createLivenessSession(
 *     deviceCorrelationId = Settings.Secure.getString(
 *         context.contentResolver,
 *         Settings.Secure.ANDROID_ID
 *     )
 * )
 *
 * // Initialize Azure Face SDK with the auth token
 * // val faceAnalyzer = FaceAnalyzerBuilder()
 * //     .setSessionAuthToken(azureSession.authToken)
 * //     .build()
 *
 * // Present liveness check UI and get result
 * // ... Azure SDK liveness check ...
 *
 * // Complete the flow
 * val result = client.completeLiveness(azureSession.azureSessionId)
 *
 * if (result.livenessResult == "live") {
 *     println("Applicant verified! ID: ${result.applicantId}")
 * }
 * ```
 */
suspend fun LivenessClient.performAzureLivenessCheck(
    applicantData: ApplicantData,
    deviceId: String
): PipelineResult = withContext(Dispatchers.IO) {
    // Step 1: Start pipeline session
    startSession(applicantData)

    // Step 2: Create Azure liveness session
    val azureSession = createLivenessSession(deviceCorrelationId = deviceId)

    // Step 3: Run Azure Face SDK liveness check
    // Note: Implement actual Azure SDK integration here
    // val faceAnalyzer = FaceAnalyzerBuilder()
    //     .setSessionAuthToken(azureSession.authToken)
    //     .build()
    // val livenessResult = faceAnalyzer.analyzeFace(activity)

    // Placeholder for SDK integration
    val sdkCompleted = true

    if (!sdkCompleted) {
        throw LivenessClientException.LivenessCheckFailed("SDK check cancelled")
    }

    // Step 4: Complete the pipeline
    val result = completeLiveness(azureSession.azureSessionId)

    val livenessResult = LivenessResult.fromString(result.livenessResult)

    PipelineResult(
        sessionId = result.sessionId,
        livenessResult = livenessResult,
        livenessConfidence = result.livenessConfidence ?: 0.0,
        applicantId = result.applicantId,
        isSuccess = livenessResult == LivenessResult.LIVE && result.applicantId != null,
        error = result.error
    )
}

File: src/mobile-client/ios/LivenessClient.swift
/**
 * iOS Client SDK for Liveness Detection Pipeline
 *
 * Integrates with the Azure-based liveness detection pipeline.
 * Uses Azure AI Vision Face SDK for client-side liveness detection.
 *
 * Prerequisites:
 * - Azure AI Vision Face SDK (AzureAIVisionFace)
 * - iOS 14.0+
 */

import Foundation
import UIKit
// Import Azure Face SDK
// import AzureAIVisionFace
// import AzureAIVisionCore

// MARK: - Models

/// Pipeline session response
public struct PipelineSession: Codable {
    public let sessionId: String
    public let status: String
    public let token: String
    public let tokenExpiresAt: String
    public let livenessResult: String?
    public let livenessConfidence: Double?
    public let applicantId: String?
    public let error: String?

    enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case status
        case token
        case tokenExpiresAt = "token_expires_at"
        case livenessResult = "liveness_result"
        case livenessConfidence = "liveness_confidence"
        case applicantId = "applicant_id"
        case error
    }
}

/// Azure liveness session for SDK
public struct AzureLivenessSession: Codable {
    public let azureSessionId: String
    public let authToken: String
    public let pipelineSessionId: String
    public let pipelineToken: String

    enum CodingKeys: String, CodingKey {
        case azureSessionId = "azure_session_id"
        case authToken = "auth_token"
        case pipelineSessionId = "pipeline_session_id"
        case pipelineToken = "pipeline_token"
    }
}

/// Applicant data for submission
public struct ApplicantData: Codable {
    public var firstName: String?
    public var lastName: String?
    public var email: String?
    public var phone: String?
    public var dateOfBirth: String?
    public var address: String?
    public var customFields: [String: String]?

    public init(
        firstName: String? = nil,
        lastName: String? = nil,
        email: String? = nil,
        phone: String? = nil,
        dateOfBirth: String? = nil,
        address: String? = nil,
        customFields: [String: String]? = nil
    ) {
        self.firstName = firstName
        self.lastName = lastName
        self.email = email
        self.phone = phone
        self.dateOfBirth = dateOfBirth
        self.address = address
        self.customFields = customFields
    }

    enum CodingKeys: String, CodingKey {
        case firstName = "first_name"
        case lastName = "last_name"
        case email
        case phone
        case dateOfBirth = "date_of_birth"
        case address
        case customFields = "custom_fields"
    }
}

/// Session creation request
public struct CreateSessionRequest: Codable {
    public let clientId: String
    public let applicantData: ApplicantData
    public var callbackUrl: String?
    public var metadata: [String: String]?

    public init(
        clientId: String,
        applicantData: ApplicantData,
        callbackUrl: String? = nil,
        metadata: [String: String]? = nil
    ) {
        self.clientId = clientId
        self.applicantData = applicantData
        self.callbackUrl = callbackUrl
        self.metadata = metadata
    }

    enum CodingKeys: String, CodingKey {
        case clientId = "client_id"
        case applicantData = "applicant_data"
        case callbackUrl = "callback_url"
        case metadata
    }
}

/// Liveness detection result
public enum LivenessResult: String {
    case live = "live"
    case spoof = "spoof"
    case uncertain = "uncertain"
    case unknown

    public init(rawValue: String) {
        switch rawValue.lowercased() {
        case "live": self = .live
        case "spoof": self = .spoof
        case "uncertain": self = .uncertain
        default: self = .unknown
        }
    }
}

/// Pipeline completion result
public struct PipelineResult {
    public let sessionId: String
    public let livenessResult: LivenessResult
    public let livenessConfidence: Double
    public let applicantId: String?
    public let isSuccess: Bool
    public let error: String?
}

// MARK: - Errors

public enum LivenessClientError: Error, LocalizedError {
    case invalidConfiguration
    case networkError(Error)
    case invalidResponse
    case sessionExpired
    case livenessCheckFailed(String)
    case serverError(String)

    public var errorDescription: String? {
        switch self {
        case .invalidConfiguration:
            return "Invalid client configuration"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .invalidResponse:
            return "Invalid server response"
        case .sessionExpired:
            return "Session has expired"
        case .livenessCheckFailed(let reason):
            return "Liveness check failed: \(reason)"
        case .serverError(let message):
            return "Server error: \(message)"
        }
    }
}

// MARK: - Client Configuration

public struct LivenessClientConfig {
    public let baseUrl: String
    public let clientId: String
    public var functionKey: String?
    public var timeout: TimeInterval = 30

    public init(
        baseUrl: String,
        clientId: String,
        functionKey: String? = nil,
        timeout: TimeInterval = 30
    ) {
        self.baseUrl = baseUrl
        self.clientId = clientId
        self.functionKey = functionKey
        self.timeout = timeout
    }
}

// MARK: - Liveness Client

/// Main client for interacting with the liveness detection pipeline
public class LivenessClient {

    private let config: LivenessClientConfig
    private let urlSession: URLSession
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    /// Current active session
    private(set) public var currentSession: PipelineSession?

    /// Initialize the client with configuration
    public init(config: LivenessClientConfig) {
        self.config = config

        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = config.timeout
        self.urlSession = URLSession(configuration: sessionConfig)
    }

    // MARK: - Session Management

    /// Start a new liveness detection session
    public func startSession(
        applicantData: ApplicantData,
        callbackUrl: String? = nil,
        metadata: [String: String]? = nil
    ) async throws -> PipelineSession {
        let request = CreateSessionRequest(
            clientId: config.clientId,
            applicantData: applicantData,
            callbackUrl: callbackUrl,
            metadata: metadata
        )

        let session: PipelineSession = try await post(
            path: "/api/sessions",
            body: request
        )

        self.currentSession = session
        return session
    }

    /// Create Azure liveness session for SDK integration
    public func createLivenessSession(
        deviceCorrelationId: String? = nil
    ) async throws -> AzureLivenessSession {
        guard let session = currentSession else {
            throw LivenessClientError.invalidConfiguration
        }

        let body: [String: String] = deviceCorrelationId.map {
            ["device_correlation_id": $0]
        } ?? [:]

        return try await post(
            path: "/api/sessions/\(session.sessionId)/liveness",
            body: body,
            token: session.token
        )
    }

    /// Complete the liveness flow after Azure SDK finishes
    public func completeLiveness(
        azureSessionId: String
    ) async throws -> PipelineSession {
        guard let session = currentSession else {
            throw LivenessClientError.invalidConfiguration
        }

        let body = ["azure_session_id": azureSessionId]

        let result: PipelineSession = try await post(
            path: "/api/sessions/\(session.sessionId)/complete",
            body: body,
            token: session.token
        )

        self.currentSession = result
        return result
    }

    /// Submit image for server-side liveness detection
    public func submitImage(_ image: UIImage) async throws -> PipelineSession {
        guard let session = currentSession else {
            throw LivenessClientError.invalidConfiguration
        }

        guard let imageData = image.jpegData(compressionQuality: 0.9) else {
            throw LivenessClientError.invalidResponse
        }

        let result: PipelineSession = try await postImage(
            path: "/api/sessions/\(session.sessionId)/image",
            imageData: imageData,
            token: session.token
        )

        self.currentSession = result
        return result
    }

    /// Get current session status
    public func getSessionStatus() async throws -> PipelineSession {
        guard let session = currentSession else {
            throw LivenessClientError.invalidConfiguration
        }

        let result: PipelineSession = try await get(
            path: "/api/sessions/\(session.sessionId)",
            token: session.token
        )

        self.currentSession = result
        return result
    }

    /// Cancel the current session
    public func cancelSession() async throws {
        guard let session = currentSession else {
            return
        }

        try await delete(
            path: "/api/sessions/\(session.sessionId)",
            token: session.token
        )

        self.currentSession = nil
    }

    // MARK: - Complete Flow Helper

    /// Execute the complete liveness detection flow
    /// This is a convenience method that handles the entire pipeline
    public func performLivenessCheck(
        applicantData: ApplicantData,
        image: UIImage
    ) async throws -> PipelineResult {
        // Step 1: Start session
        let session = try await startSession(applicantData: applicantData)

        // Step 2: Submit image for liveness detection
        let result = try await submitImage(image)

        // Step 3: Build result
        let livenessResult = LivenessResult(rawValue: result.livenessResult ?? "unknown")

        return PipelineResult(
            sessionId: result.sessionId,
            livenessResult: livenessResult,
            livenessConfidence: result.livenessConfidence ?? 0,
            applicantId: result.applicantId,
            isSuccess: livenessResult == .live && result.applicantId != nil,
            error: result.error
        )
    }

    // MARK: - Network Helpers

    private func buildURL(path: String) -> URL? {
        let baseUrl = config.baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return URL(string: baseUrl + path)
    }

    private func buildRequest(
        url: URL,
        method: String,
        token: String? = nil
    ) -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = method

        if let key = config.functionKey {
            request.setValue(key, forHTTPHeaderField: "x-functions-key")
        }

        if let token = token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        return request
    }

    private func get<T: Decodable>(
        path: String,
        token: String? = nil
    ) async throws -> T {
        guard let url = buildURL(path: path) else {
            throw LivenessClientError.invalidConfiguration
        }

        let request = buildRequest(url: url, method: "GET", token: token)

        return try await execute(request)
    }

    private func post<T: Decodable, B: Encodable>(
        path: String,
        body: B,
        token: String? = nil
    ) async throws -> T {
        guard let url = buildURL(path: path) else {
            throw LivenessClientError.invalidConfiguration
        }

        var request = buildRequest(url: url, method: "POST", token: token)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(body)

        return try await execute(request)
    }

    private func postImage<T: Decodable>(
        path: String,
        imageData: Data,
        token: String? = nil
    ) async throws -> T {
        guard let url = buildURL(path: path) else {
            throw LivenessClientError.invalidConfiguration
        }

        var request = buildRequest(url: url, method: "POST", token: token)
        request.setValue("image/jpeg", forHTTPHeaderField: "Content-Type")
        request.httpBody = imageData

        return try await execute(request)
    }

    private func delete(
        path: String,
        token: String? = nil
    ) async throws {
        guard let url = buildURL(path: path) else {
            throw LivenessClientError.invalidConfiguration
        }

        let request = buildRequest(url: url, method: "DELETE", token: token)

        do {
            let (_, response) = try await urlSession.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw LivenessClientError.invalidResponse
            }

            if httpResponse.statusCode >= 400 {
                throw LivenessClientError.serverError("HTTP \(httpResponse.statusCode)")
            }
        } catch let error as LivenessClientError {
            throw error
        } catch {
            throw LivenessClientError.networkError(error)
        }
    }

    private func execute<T: Decodable>(_ request: URLRequest) async throws -> T {
        do {
            let (data, response) = try await urlSession.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw LivenessClientError.invalidResponse
            }

            if httpResponse.statusCode == 401 {
                throw LivenessClientError.sessionExpired
            }

            if httpResponse.statusCode >= 400 {
                if let errorData = try? decoder.decode([String: String].self, from: data),
                   let error = errorData["error"] {
                    throw LivenessClientError.serverError(error)
                }
                throw LivenessClientError.serverError("HTTP \(httpResponse.statusCode)")
            }

            return try decoder.decode(T.self, from: data)
        } catch let error as LivenessClientError {
            throw error
        } catch let error as DecodingError {
            print("Decoding error: \(error)")
            throw LivenessClientError.invalidResponse
        } catch {
            throw LivenessClientError.networkError(error)
        }
    }
}

// MARK: - Azure SDK Integration

/**
 * Integration with Azure AI Vision Face SDK
 *
 * Example usage:
 *
 * ```swift
 * let client = LivenessClient(config: LivenessClientConfig(
 *     baseUrl: "https://your-function-app.azurewebsites.net",
 *     clientId: "your-client-id",
 *     functionKey: "your-function-key"
 * ))
 *
 * // Start session
 * let session = try await client.startSession(applicantData: ApplicantData(
 *     firstName: "John",
 *     lastName: "Doe",
 *     email: "john@example.com"
 * ))
 *
 * // Get Azure liveness session
 * let azureSession = try await client.createLivenessSession()
 *
 * // Initialize Azure Face SDK with the auth token
 * // let faceAnalyzer = try FaceAnalyzerBuilder()
 * //     .sessionAuthToken(azureSession.authToken)
 * //     .build()
 *
 * // Present liveness check UI and get result
 * // ... Azure SDK liveness check ...
 *
 * // Complete the flow
 * let result = try await client.completeLiveness(
 *     azureSessionId: azureSession.azureSessionId
 * )
 *
 * if result.livenessResult == "live" {
 *     print("Applicant verified! ID: \(result.applicantId ?? "N/A")")
 * }
 * ```
 */
extension LivenessClient {

    /// Execute liveness check using Azure Face SDK
    /// This method integrates with Azure AI Vision Face SDK for client-side liveness
    public func performAzureLivenessCheck(
        applicantData: ApplicantData,
        presentingViewController: UIViewController
    ) async throws -> PipelineResult {
        // Step 1: Start pipeline session
        let _ = try await startSession(applicantData: applicantData)

        // Step 2: Create Azure liveness session
        let azureSession = try await createLivenessSession(
            deviceCorrelationId: UIDevice.current.identifierForVendor?.uuidString
        )

        // Step 3: Run Azure Face SDK liveness check
        // Note: Uncomment and implement when Azure Face SDK is added
        /*
        let faceAnalyzer = try FaceAnalyzerBuilder()
            .sessionAuthToken(azureSession.authToken)
            .build()

        let livenessResult = try await faceAnalyzer.analyzeFace(
            from: presentingViewController
        )
        */

        // Placeholder for SDK integration
        // In production, replace with actual Azure SDK call
        let sdkCompleted = true

        guard sdkCompleted else {
            throw LivenessClientError.livenessCheckFailed("SDK check cancelled")
        }

        // Step 4: Complete the pipeline
        let result = try await completeLiveness(
            azureSessionId: azureSession.azureSessionId
        )

        let livenessResult = LivenessResult(rawValue: result.livenessResult ?? "unknown")

        return PipelineResult(
            sessionId: result.sessionId,
            livenessResult: livenessResult,
            livenessConfidence: result.livenessConfidence ?? 0,
            applicantId: result.applicantId,
            isSuccess: livenessResult == .live && result.applicantId != nil,
            error: result.error
        )
    }
}
